packages/ragbits-core/src/ragbits/core/vector_stores/pgvector.py

```diff
@@ -0,0 +1,334 @@
+
+        Args:
+            client: The pgVector database connection pool.
+            table_name: The name of the table.
+            vector_size: The size of the vectors.
+            distance_method: The distance method to use.
+            hnsw_params: The parameters for the HNSW index. If None, the default parameters will be used.
+            default_options: The default options for querying the vector store.
+            metadata_store: The metadata store to use. If None, the metadata will be stored in pgVector db.
+        """
+        super().__init__(default_options=default_options, metadata_store=metadata_store)
+
+        if not re.match(r"^[a-zA-Z_][a-zA-Z0-9_]*$", table_name):
+            raise ValueError(f"Invalid table name: {table_name}")
+
+        if hnsw_params is None:
+            hnsw_params = {"m": 4, "ef_construction": 10}
+        self._client = client
+        self._table_name = table_name
+        self._vector_size = vector_size
+        self._distance_method = distance_method
+        self._hnsw_params = hnsw_params
+
+    def _create_table_command(self) -> str:
+        """
+        Create sql query for creating a pgVector table.
+
+        Returns:
+              str: sql query.
+        """
+        type_mapping = {
+            str: "TEXT",
+            list: f"VECTOR({self._vector_size})",
+            dict: "JSONB",
+        }
+        columns = []
+        type_hints = get_type_hints(VectorStoreEntry)
+        for column, column_type in type_hints.items():
+            if column_type == list[float]:
+                columns.append(f"{column} {type_mapping[list]}")
+            else:
+                sql_type = type_mapping.get(column_type)
+                columns.append(f"{column} {sql_type}")
+
+        return f"CREATE TABLE {self._table_name} (" + ", ".join(columns) + ");"
+
+    def _create_retrieve_query(
+        self, vector: list[float], query_options: VectorStoreOptions | None = None
+    ) -> tuple[str, list[Any]]:
+        """
+        Create sql query for retrieving entries from the pgVector collection.
+
+        Args:
+            vector: The vector to query.
+            query_options: The options for querying the vector store.
+
+        Returns:
+            str: sql query.
+        """
+        distance_operator = DISTANCE_OPS[self._distance_method][1]
+        if not query_options:
+            query_options = self.default_options
+
+        # _table_name has been validated in the class constructor, and it is a valid table name.
+        query = f"SELECT * FROM {self._table_name}"  # noqa S608
+
+        values = []
+        index = 1
+
+        if query_options.max_distance and self._distance_method == "ip":
+            query += f""" WHERE vector ${index} '${index + 1}'
+            BETWEEN ${index + 2} AND ${index + 3}"""
+            values.extend([distance_operator, vector, (-1) * query_options.max_distance, query_options.max_distance])
+            index += 4
```

Comment:
nit: I would probably based those calculation on the length of `values` instead of keeping a separate `index` - less chance that the two desynchronize. But that's a personal preference.